jctools-core/src/main/java/org/jctools/queues/MpscBlockingConsumerArrayQueue.java

```diff
@@ -428,32 +424,22 @@ public E take() throws InterruptedException
             final long pIndex = lvProducerIndex();
             if (cIndex == pIndex && casProducerIndex(pIndex, pIndex + 1))
             {
-                boolean unblocked = false;
-                try
+                // producers only try a wakeup when both the index and the blocked thread are visible
+                soBlocked(Thread.currentThread());
+                do
                 {
-                    // producers only try a wakeup when both the index and the blocked thread are visible
-                    soBlocked(Thread.currentThread());
-                    do
+                    LockSupport.park();
+                    if (Thread.interrupted())
                     {
-                        LockSupport.park();
-                        if (Thread.interrupted())
-                        {
-                            throw new InterruptedException();
-                        }
-                    }
-                    while (lvBlocked() != null);
-                    unblocked = true;
-                }
-                finally
-                {
-                    if (!unblocked) {
                         // revert blocking state
                         if (casProducerIndex(pIndex + 1, pIndex))
                         {
                             soBlocked(null);
                         }
+                        throw new InterruptedException();
```

Comment:
@philipa I think the cas-fail case should be handled differently here - to basically just ignore the interruption but don't reset the interrupt status then continue spinning for unblock and then return the element instead of throwing, like I did in the `take(timeout)` case below. Otherwise this element will be "lost"

Reply:
How is the element lost? We've not advanced `cIndex` and the producer is in control of `pIndex`.

Reply:
@philipa oops yes you're right, my bad... so then it would make sense for the interrupt/timeout logic in `take(timeout)` to also match this for simplicity.

However I think you do still need to spin-wait to see `unblocked` null prior to returning/throwing in all these cas failure cases or else there is still a race possible.

Reply:
I don't think the spin-wait is required. The producer has the responsibility of unblocking and cleaning up. Isn't identical handling sufficient?

                    if (Thread.interrupted())
                    {
                        if (casProducerIndex(pIndex + 1, pIndex)) {
                            soBlocked(null);
                        }
                        throw new InterruptedException();
                    }

Reply:
@philipa what about:

C: interrupted in `take()`
P1: offers, wins producer index race, `soRefElement(A)`
C: throws IE
C: `take()` again consumes `A`
C: `get(timeout)`, cas `soBlocked(currentThread)`, enter park loop
P1: soBlocked(null), unpark
C: breaks from loop, enters `spinWaitForElement`
timeout elapses or C is interrupted, C does not return (at least until further producer activity)

Reply:
I had to stare at this, but .... agreed. It's similar to the original race. It could also happen using `take()` throughout.

So the consumer should always `spinWaitForUnblocked()` if it fails to revert the blocking state. I'll do that.